### ISSUE TITLE
PinPoint Analytics to Cognito Application Client not provisioned correctly

### ISSUE THREAD

**Issue by whatnick** [hashicorp/terraform-provider-aws#16481]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

```
Terraform v0.13.5
+ provider registry.terraform.io/hashicorp/aws v3.18.0
```

### Affected Resource(s)

* aws_cognito_user_pool_client

### Terraform Configuration Files

Cognito module - https://github.com/opendatacube/datacube-k8s-eks/tree/master/cognito

### Debug Output

None ( works without failure but resources provisioned not as expected)

### Panic Output

None

### Expected Behavior
PinPoint projects per Cognito application client produced with correct linkages.

### Actual Behavior

TerraForm provisioned PinPoint resources without linking them correctly to Cogntion Application Clients.

![image](https://user-images.githubusercontent.com/491396/100560360-ab80cf80-3309-11eb-9b2b-daeae84f5689.png)

### Steps to Reproduce

Create a Cognito application client using the module linked above and set `enable_pinpoint=true`

1. `terraform apply`

### References

* Part of JupyterHub + EKS + Cognito analytics work - https://github.com/opendatacube/datacube-k8s-eks/issues/251

**Comment by DrFaust92** [comment]
from a quick look in looks like the role used for cognito to call pinpoint is missing a condition for `sts:ExternalId`

you should have the following your assume role policy:

```
      Action = "sts:AssumeRole"
      Condition = {
        "StringEquals" = {
          "sts:ExternalId" = {external_id_used_in_config}
        }
      }
```

and as i see the external_id argument is the key from the for each so will either need a separate role per client or ditch the `external_id` argument (albeit it makes it less secure technically)

**Comment by whatnick** [comment]
Experimenting with the easy / insecure externalID wildcard gives the same unlinked results. 
![image](https://user-images.githubusercontent.com/491396/100832506-212c9d00-34bc-11eb-8535-0e3fd8ad90e6.png)

In contract the Role generated by AWS Console does not have the ExternalID condition.
![image](https://user-images.githubusercontent.com/491396/100832642-618c1b00-34bc-11eb-9f68-b8bbaa65ad33.png)

**Comment by DrFaust92** [comment]
Ok, so i checked this and something is wrong on AWS side. when i try this on us-east-1 the application shows up fine but for other regions (us-west-2) it doesn't pop up in the console but the API does return the correct response (and thats why terraform doesn't error or crash).

I tried it both with the external id condition and without but it worked fine with both. another thing is the drop down for choosing an application is empty even when applications exist.
unfortunately what i can only recommend at this point is to reach out to AWS and debug this with them. optionally move to this application to us-east-1 where it seems to work fine.

**Comment by whatnick** [comment]
I have this from AWS support
 ```We’ve recently updated our Pinpoint parameter set when moving to in-region integrations from us-east-1 only. To point to SYD in-region, please use the following new parameter applicationArn (leaving out roleArn/externalID) like so:

aws cognito-idp create-user-pool-client --user-pool-id <user-pool-id> --client-name pinpoint_test_client --analytics-configuration ApplicationArn=arn:aws:mobiletargeting:ap-southeast-2:<your-account-id>:apps/<your-syd-pinpoint-app-id>,UserDataShared=true --region ap-southeast-2

Previously, Cognito only supported sending events to us-east-1 Pinpoint projects. When we moved towards supporting in-region Pinpoint, we also started using Service Linked Roles to support the integrations instead of a user provided role or service-role. Supporting customer’s existing us-east-1 Pinpoint projects as well as new in-region projects, we eliminated the ambiguity and introduced a new cli parameter applicationArn. Specifying this parameter allows in-region and also uses the SLR for the Pinpoint integration.
As to not break customers, we still allow them to provide their own role to publish events via the roleArn/externalID parameters, but this legacy method will continue to use Pinpoint in us-east-1 only. Please use applicationArn to publish events in-region where available. We will update the documentation to clear the confusion surrounding in-region integrations.```

**Comment by DrFaust92** [comment]
Hey @whatnick ill take a look if this is implemented in the sdk and test it. will update.
If i find a way to make this work ill create a PR to deal with this.

Thanks for bringing this up and working with support to debug this!

**Comment by DrFaust92** [comment]
Opened https://github.com/hashicorp/terraform-provider-aws/pull/16734 to address this.

**Comment by bflad** [comment]
Support for the new `application_arn` argument has been merged and will release with version 3.29.0 of the Terraform AWS Provider, later today. Thank you to @DrFaust92 for the implementation. 👍
